sdk/core/azure-core/azure/core/rest/_http_response_impl.py

```diff
@@ -0,0 +1,320 @@
+        if not self._json:
+            self._json = loads(self.text())
+        return self._json
+
+    def _stream_download_check(self):
+        if self.is_stream_consumed:
+            raise StreamConsumedError(self)
+        if self.is_closed:
+            raise StreamClosedError(self)
+
+        self._is_stream_consumed = True
+
+    def raise_for_status(self):
+        # type: (...) -> None
+        """Raises an HttpResponseError if the response has an error status code.
+
+        If response is good, does nothing.
+        """
+        if cast(int, self.status_code) >= 400:
+            raise HttpResponseError(response=self)
+
+    @property
+    def content(self):
+        # type: (...) -> bytes
+        """Return the response's content in bytes."""
+        if self._content is None:
+            raise ResponseNotReadError(self)
+        return self._content
+
+    def __repr__(self):
+        # type: (...) -> str
+        content_type_str = (
+            ", Content-Type: {}".format(self.content_type) if self.content_type else ""
+        )
+        return "<HttpResponse: {} {}{}>".format(
+            self.status_code, self.reason, content_type_str
+        )
+
+class HttpResponseImpl(_HttpResponseBaseImpl, _HttpResponse):
+    """HttpResponseImpl built on top of our HttpResponse protocol class.
+
+    Since ~azure.core.rest.HttpResponse is an abstract base class, we need to
+    implement HttpResponse for each of our transports. This is an implementation
+    that each of the sync transport responses can be built on.
+
+    :keyword request: The request that led to the response
+    :type request: ~azure.core.rest.HttpRequest
+    :keyword any internal_response: The response we get directly from the transport. For example, for our requests
+     transport, this will be a requests.Response.
+    :keyword optional[int] block_size: The block size we are using in our transport
+    :keyword int status_code: The status code of the response
+    :keyword str reason: The HTTP reason
+    :keyword str content_type: The content type of the response
+    :keyword MutableMapping[str, str] headers: The response headers
+    :keyword Callable stream_download_generator: The stream download generator that we use to stream the response.
+    """
+
+    def __enter__(self):
+        # type: (...) -> HttpResponseImpl
+        return self
+
+    def close(self):
+        # type: (...) -> None
+        if not self.is_closed:
+            self._is_closed = True
+            self._internal_response.close()
+
+    def __exit__(self, *args):
+        # type: (...) -> None
+        self.close()
+
+    def _set_read_checks(self):
+        self._is_stream_consumed = True
+        self.close()
+
+    def read(self):
+        # type: (...) -> bytes
+        """
+        Read the response's bytes.
+
+        """
+        if self._content is None:
+            self._content = b"".join(self.iter_bytes())
+        self._set_read_checks()
```

Comment:
Is this by design that read will close the response?

Reply:
yeah, since in the current impl, the transport returns the old response, and we have to convert from the old to the new, the body has already been read in. once we switch the pipelines over to the new responses, us reading the new body will automatically close it, but for now, we have to patch a little bit

Reply:
Then we need to re-scan our code. I see several places that we read and close. :)

Reply:
Still, we have close() in both iter_bytes and _set_read_checks?

Reply:
I'll get rid of all of them, apologies for not doing sooner!

Reply:
why not move this into if?